### PR TITLE
341 Manage LLMs (II)

### PR DESCRIPTION
**Pull request by stephan-buckmaster**
The continue https://github.com/AllYourBot/hostedgpt/pull/385

### CODE REVIEW

**Review by krschacht — CHANGES_REQUESTED** [review]
I just did a quick skim through this and flagged a few small items that I noticed.

To reply to your other comment:

Thanks for the feedback.

> It's a good point that it looks heavy, that language models and API services are separated. Indeed I had them combined initially. Then I looked at the database schema and it wasn't normalized; then I used the app the next day and it was annoying to look up the port and access token. So once one has a API server setup I think it's natural to have that reflected in one database table, and one UI section. Its often effort to locate the access-token (again)

I also agree they should be separate. My comment was slightly different. I was proposing we do keep them the same, but currently you edit an Assistant and you assign a LanguageModel and an API. I was proposing instead that from Assistant you just assign a LanguageModel, and then you edit a LanguageModel to assign it to an API. But APIs would still be their own entities and they would each of their own section within the Settings and you could still assign multiple LanguageModels to the same API so you wouldn’t have to keep specifying the port & token repeatedly.

> I feel the assistants should not be as prominent, get that much space.

Yea, after I saw your two new sections it made me realize that Assistants should just be a top-level menu item and then clicking that would list the individual assistants. Eventually it won’t make sense to show all the assistants in the left side, in the settings.

In terms of the display you suggest, I'll work on item 1,3.

> Item 2: So you're saying instead of cards, just have in effect, a paragraph, per entry. Then an Edit link if applicable.

I was proposing to just have a row per entry and then either an “Edit” link on the row or maybe the whole row is just clickable.

> Item 4: The "system" entries (have user_id=NULL) wouldn't be editable by any user, in my mind. So they wouldn't get shown in a form.

Oh, yes. Good point. I guess I would propose then: when you click on a system entry, it’s a show page (no form, no edit link anywhere). And when you click on a user one then it takes you to the form, and clicking save on the form just refreshes and keeps you on the form. I don’t think we need to toggle back & forth between an edit & show for user ones.

The one issue I’m aware of with this approach is that the top-right corner toast notification which says “Saved” is not quite prominent enough. I’ve observed people click save and not realize it saved. I plan to move that to the center of the screen and have it slide in a more visually prominent way.

File: app/models/user.rb [hunk 1]
```diff
@@ -13,6 +13,10 @@ class User < ApplicationRecord
 
   has_many :assistants, -> { not_deleted }
   has_many :assistants_including_deleted, class_name: "Assistant", dependent: :destroy
+  has_many :language_models, -> { not_deleted }
+  has_many :language_models_including_deleted, class_name: "APIService", dependent: :destroy
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/user.rb:17]: Fun, thanks for that

File: app/models/api_service.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+class APIService < ApplicationRecord
+  encrypts :access_token
+
+  DRIVERS = %w(Anthropic OpenAI)
+
+  belongs_to :user
+  validates :url, format: URI::DEFAULT_PARSER.make_regexp(%w[http https]),  if: -> { url.present? }
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/api_service.rb:7]: The next line has the "presence" validation :
`validates :name, :url, presence: true`

[the remark this answers is not in the document]
Comment (krschacht) [reply app/models/api_service.rb:7]: Ah, yes, then you should surely be able to remove `if: -> { url.present? }` from here, right? It won't even let you save blank ones

**Comment by krschacht** [comment]
@stephan-buckmaster Recent changes of mine created some new conflicts on this branch. Let me know if you want help resolving those; I can take a pass at it. I was just about to do it but I stopped myself b/c I don't want to push a commit to your branch and risk colliding with you. Just let me know if you want help with those conflicts.

File: app/models/api_service.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/api_service.rb:7]: It's actually the other way around. When removing it, and the URL is empty, one get's two errors, one from each validation: "URL is invalid" and "URL can't be blank." But one is enough

**Comment by stephan-buckmaster** [comment]
The merge conflicts look managable. I'll let you know, @krschacht

**Comment by stephan-buckmaster** [comment]
I believe I got through all your suggestions @krschacht.

Still no tests (there's a few harmless failures).
Let me know if you see anything off, while I work on those ...

File: db/migrate/20240606012618_remove_auth_from_users.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ def up
       credential = user.credentials.build(
         type: "PasswordCredential",
         password_digest: user.password_digest,
-        last_authenticated_at: user.person.updated_at
```
Comment (stephan-buckmaster) [inline db/migrate/20240606012618_remove_auth_from_users.rb:7]: This ran into an error with my dev db, a user witout a person record, so added the rescue-nil

Comment (krschacht) [reply db/migrate/20240606012618_remove_auth_from_users.rb:7]: Hmm, that's interesting. I don't think it should be possible to have a User without a Person. If that's happening, it suggests a deeper bug in the code and possibly a validation that's missing.

I just did a big extension of auth and credentials (merged into main last week) and some of the auth-related stuff is associated with the user (like credentials) but credentials let you auth different clients (i.e. browsers) and clients are associated with the person.

Anyway, just sharing that as food for thought. It's fine to add the rescue nil in this migration anyway but I think there's a deeper problem somewhere.

Comment (stephan-buckmaster) [reply db/migrate/20240606012618_remove_auth_from_users.rb:7]: Maybe my fixtures were not in good shape

**Comment by stephan-buckmaster** [comment]
Ok, pushed rather large changes, including many tests. Could have some gaps, I'll revisit on the weekend. 

Let me know your thoughts in the mean time.

**Comment by krschacht** [comment]
Great, I'll give it a close read through. Are you okay if I commit some changes to this branch as I'm reading through it? If I notice little clean up things it's often faster for me to just make the change right then, but I don't want us to step on each others' toes. :)

**Comment by stephan-buckmaster** [comment]
Sure go ahead. (I realized I didn't work on the display of system-wide language-models, they look a bit weak.)

File: app/services/ai_backend/anthropic.rb [hunk 4]
```diff
@@ -13,9 +13,14 @@ def self.client
   end
 
   def initialize(user, assistant, conversation, message)
-    raise ::Anthropic::ConfigurationError if user.anthropic_key.blank?
     begin
-      @client = self.class.client.new(access_token: user.anthropic_key)
+      if assistant.api_service.blank?
```
Comment (krschacht) [inline app/services/ai_backend/anthropic.rb:17]: This branching doesn't feel quite right. Why are there any assistants that do not have an `api_service`? Why don't we just create api_services for all of our system assistants? If we keep this inconsistency then we'll have to add these sprinkle these conditional checks throughout the code, whereas if we resolve the inconsistency then we can always assume an assistant has an api_service

**Review by krschacht — CHANGES_REQUESTED** [review]
This is a meaty PR but it's looking really good! Don't be scared by all my comments :) the vast majority of them are really quick, really small suggestions. And I used the github "suggest changes" feature (I had forgotten about this) so most of them are just quick accepts. I only did two tiny commits that involved reordering some things within a model. They're not even worth looking at.

The only thing I haven't gone through in detail are your tests.

With regards to the UI, I think what you implemented is a good v1 so I'm fine without you doing any further visual changes. There are a couple small things I want to tweak but I'll do that in a follow-up PR. I have an idea for a new way of extracting some views things into components, and the settings area of the app is actually a nice place for me to implement it first (e.g. we have a table view in two places now and I have a clever idea for extracting that as a view component). But I don't want to add any more scope to this so I kept all my feedback restricted to actual logic of the feature.

File: db/migrate/20240531134924_add_api_service_id_to_assistants.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddAPIServiceIdToAssistants < ActiveRecord::Migration[7.1]
+  def change
+    add_column :assistants, :api_service_id, :bigint, null: true
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply db/migrate/20240531134924_add_api_service_id_to_assistants.rb]: Done

File: db/schema.rb [hunk 6]
```diff
@@ -109,6 +145,19 @@
     t.boolean "supports_images", null: false
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.bigint "user_id"
+    t.datetime "deleted_at", precision: nil
+    t.index ["user_id", "deleted_at"], name: "index_language_models_on_user_id_and_deleted_at"
+  end
+
+  create_table "memories", force: :cascade do |t|
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply db/schema.rb]: Should be good

File: db/migrate/20240526195302_expand_language_models.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class ExpandLanguageModels < ActiveRecord::Migration[7.1]
+  def change
+    add_column :language_models, :user_id, :bigint, null: true
+    add_column :language_models, :deleted_at, :timestamp, default: nil
+
+    add_index :language_models, [:user_id, :deleted_at]
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply db/migrate/20240526195302_expand_language_models.rb:6]: Done

File: db/migrate/20240530162948_create_api_services.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+class CreateAPIServices < ActiveRecord::Migration[7.1]
+  def change
+    create_table :api_services do |t|
+      t.bigint :user_id
+      t.string :name, null: false
+      t.string :driver, null: false
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply db/migrate/20240530162948_create_api_services.rb:6]: Done

File: db/migrate/20240530162948_create_api_services.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+class CreateAPIServices < ActiveRecord::Migration[7.1]
+  def change
+    create_table :api_services do |t|
+      t.bigint :user_id
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply db/migrate/20240530162948_create_api_services.rb:4]: Done

File: db/migrate/20240530162948_create_api_services.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+      t.string :name, null: false
+      t.string :driver, null: false
+      t.string :url, null: false
+      t.string :access_token, null: true
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply db/migrate/20240530162948_create_api_services.rb:8]: Done

[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply db/migrate/20240530162948_create_api_services.rb:8]: Made change, although one gets code like this,  `access_token: assistant.api_service.token` when invoking the backend.

File: app/controllers/settings/api_services_controller.rb [hunk 11]
```diff
@@ -0,0 +1,57 @@
+class Settings::APIServicesController < Settings::ApplicationController
+  before_action :set_api_service, only: [:edit, :update, :destroy]
+
+  def index
+    @api_services = Current.user.api_services.all.order(updated_at: :desc)
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/controllers/settings/api_services_controller.rb:5]: Done

File: app/controllers/settings/api_services_controller.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+  def new
+    @api_service = APIService.new
+  end
+
+  def create
+    @api_service = Current.user.api_services.new(api_service_params)
+
+    if @api_service.save
+      redirect_to settings_api_services_path, notice: "Saved"
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/controllers/settings/api_services_controller.rb:19]: Done

File: app/controllers/settings/api_services_controller.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  private
+
+  def set_users_api_service
+    @api_service = Current.user.api_services.find_by(id: params[:id])
+    if @api_service.nil?
+      redirect_to new_settings_api_service_url, notice: "The API Service was deleted", status: :see_other
+    end
+  end
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/controllers/settings/api_services_controller.rb:45]: Removed

File: app/models/language_model.rb [hunk 14]
```diff
@@ -1,27 +1,60 @@
...
+    end
+  end
 
   def readonly?
-    !new_record?
+    user_id.blank?
+  end
+
+  def destroy
+    raise ActiveRecord::ReadOnlyError 'System model cannot be deleted' if user.blank?
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/language_model.rb:36]: I think this is the right place. There is a delete button for language models

File: app/services/ai_backend/anthropic.rb (hunk 4 above)
Comment (stephan-buckmaster) [reply app/services/ai_backend/anthropic.rb:17]: Not sure, it seems to me then we'll have the same thing with the OpenAI/Anthropic URL; we don't want to configure those URLs in our database do we?

File: app/controllers/settings/language_models_controller.rb [hunk 15]
```diff
@@ -0,0 +1,58 @@
+class Settings::LanguageModelsController < Settings::ApplicationController
+  before_action :set_users_language_model, only: [:edit, :update, :destroy]
+  before_action :set_language_model, only: [:show]
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/controllers/settings/language_models_controller.rb:3]: Done

File: app/controllers/settings/language_models_controller.rb [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+  def set_users_language_model
+    @language_model = Current.user.language_models.find_by(id: params[:id])
+    if @language_model.nil?
+      redirect_to new_settings_language_model_url, notice: "The Language Model could not be found", status: :see_other
+    end
+  end
+
+  def set_language_model
+    @language_model = LanguageModel.find_by(id: params[:id])
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/controllers/settings/language_models_controller.rb:52]: There was a test "should not update other user's language_model" for the controller, I added now also,  "should not allow viewing other user's records" So I think no code change is needed.

File: app/views/settings/language_models/_form.html.erb [hunk 17]
```diff
@@ -0,0 +1,58 @@
+<%= form_with(model: [:settings, language_model], class: "contents") do |form| %>
+  <% if language_model.errors.any? %>
+    <div id="error_explanation" class="bg-red-50 text-red-500 px-3 py-2 font-medium rounded-lg mt-3">
+      <h2><%= pluralize(language_model.errors.count, "error") %> prohibited this language_model from being saved:</h2>
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/views/settings/language_models/_form.html.erb:4]: Done

File: app/views/settings/language_models/edit.html.erb [hunk 18]
```diff
@@ -0,0 +1,25 @@
+<div class="mx-auto md:w-2/3 w-full">
+  <h1 class="font-bold text-4xl">
+   Editing  Language Model <%= @language_model.name %>
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/views/settings/language_models/edit.html.erb:3]: Done

File: app/views/settings/api_services/_form.html.erb [hunk 19]
```diff
@@ -0,0 +1,55 @@
+<%= form_with(model: [:settings, api_service], class: "contents") do |form| %>
+  <% if api_service.errors.any? %>
+    <div id="error_explanation" class="bg-red-50 text-red-500 px-3 py-2 font-medium rounded-lg mt-3">
+      <h2><%= pluralize(api_service.errors.count, "error") %> prohibited this api_service from being saved:</h2>
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/views/settings/api_services/_form.html.erb:4]: Done

File: app/models/api_service.rb [hunk 20]
```diff
@@ -0,0 +1,31 @@
+class APIService < ApplicationRecord
+  encrypts :access_token
+
+  DRIVERS = %w(Anthropic OpenAI)
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/api_service.rb]: Got it

File: app/models/user.rb [hunk 21]
```diff
@@ -27,6 +32,24 @@ def preferences
     attributes["preferences"].with_defaults(dark_mode: "system")
   end
 
+  # Which ones can they choose for their api services
+  def usable_language_models
+    LanguageModel.for_user(self).not_deleted
+  end
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/user.rb:38]: Found no references :-)

File: app/models/user.rb [hunk 22]
```diff
@@ -27,6 +32,24 @@ def preferences
...
+
+  def destroy
+    @destroy_in_progress = true
+    begin
+      super
+    ensure
+      @destroy_in_progress = false
+    end
+  end
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/user.rb:51]: Ok I'll save this for another session, tomorrow ?

File: app/models/assistant.rb [hunk 23]
```diff
@@ -18,6 +18,26 @@ class Assistant < ApplicationRecord
 
   scope :ordered, -> { order(:id) }
 
+  delegate :api_service, to: :language_model
```
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/assistant.rb:21]: Done

**Comment by stephan-buckmaster** [comment]
I hope my comments are visible above , many "Done". Let me know about the test changes, please @krschacht . Still need to look at the delete/destroy issue you mentioned

File: app/services/ai_backend/anthropic.rb (hunk 4 above)
Comment (krschacht) [reply app/services/ai_backend/anthropic.rb:17]: I think we should configure those URLs in our database. The fact that we have two different ways of getting a URL is a good clue. If some assistant’s don’t have an api_service and others do, we’ll surely end up adding more conditionals over time to account for that difference.

If someone wanted to switch from OpenAI’s servers to using Microsoft Azure for all their GPT-4 needs, this would be the place to change it. It would be really nice to be able to edit that within the app. Also, I think we’ll inevitably end up adding additional logic to the API Service abstraction. Now that it exists, it’s a good place to house certain things. For example, one thing I’m working on is voice mode for the app. I did some digging and discovered that ChatGPT layers on some additional custom instructions when you activate the microphone. The GPT is told “the text you generate will be spoken so don’t use bullet points”.

This is not something I want to do per assistant or even per language model, I think it’s a setting which  I would add to the api_service model. I’ll pre-populate some “voice instructions” will get concatenated to your Assistant’s custom instructions, but if you want to tweak that setting it would get tweaked at this low level and apply to all assistants.

File: test/fixtures/api_services.yml [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+  url: http://taylor.org/doit
+  token: ta-secret
+  driver: Anthropic
+
+taylor_openai_service:
+  name: Taylor OpenAI Server
+  user: :taylor
+  url: http://taylor.org/doit-open
+  token: open-secret
+  driver: OpenAI
```
Comment (krschacht) [inline test/fixtures/api_services.yml:27]: I think we should put real data in here: the actual data we want for OpenAI and Anthropic. And then we should update seeds.rb to preload this file. This will ensure all new installations get the two core api_services pre-populated.

File: test/fixtures/language_models.yml [hunk 25]
```diff
@@ -2,144 +2,176 @@
...
+  description: Alpaca (large) from fixtures
+  supports_images: true
+
+pacos:
+  user: taylor
+  position: 28
+  api_name: pacos-imagine
+  api_service: taylor_openai_service
+  description: Lama pacos (large) from fixtures
+  supports_images: true
```
Comment (krschacht) [inline test/fixtures/language_models.yml:177]: Related to the comment I leave on api_service, this language_model fixture is seeded before each test, but we also use this fixture to seed new installations (seeds.rb) so everything in here will appear in production. This is because it’s really easy to keep inadvertently making code changes that break the app for new users; I’ve done it a bunch of times. My thinking has been: if I keep dev and test closer to production, then it’s easier to keep the app stable.

So basically, this new test data you added here will appear for new installs. Can you rewrite the tests to use the existing language model fixture data?

File: test/models/api_service_test.rb [hunk 26]
```diff
@@ -0,0 +1,96 @@
+require "test_helper"
+
+class APIServiceTest < ActiveSupport::TestCase
+  test "has user" do
+    assert_equal users(:keith), api_services(:keith_service).user
```
Comment (krschacht) [inline test/models/api_service_test.rb:5]: Awhile ago I updated all the associations tests to use:

```suggestion
    assert_instance_of User, api_services(:keith_service).user
```

**Comment by krschacht** [comment]
> I hope my comments are visible above , many "Done". Let me know about the test changes, please @krschacht . Still need to look at the delete/destroy issue you mentioned

Ah yes, I forgot I had not reviewed tests. I just did that now and they look good. Nice job with the test coverage. There’s a fine line between not enough testing and too much and I think you nailed it. My one comment is around keeping the fixtures “real data” for the two models we’re going to populate in production (language_model and api_service).

File: app/services/ai_backend/anthropic.rb (hunk 4 above)
Comment (stephan-buckmaster) [reply app/services/ai_backend/anthropic.rb:17]: If you're thinking about changing these URLs then they would need to be user-level. So every user will have these user-level api-services, implying the existing language_models records we created with the db-migration will be moved(copied) to user-level too then?  It does sound consistent, but I wanted to make sure I understand your suggestion right.

File: test/fixtures/api_services.yml (hunk 24 above)
Comment (stephan-buckmaster) [reply test/fixtures/api_services.yml:27]: It seems more appropriate to go to lengths to **not** get these values into the prod database. Since they are for actual users. What do I know what real-life users Keith and Rob want in their configurations. (I just made up Taylor.)

However, for the OpenAI/Anthropic api_services values, it seem they should get hard-coded in the User::Registerable.create_initial_assistants method (and db-migration for existing users)

File: test/fixtures/language_models.yml (hunk 25 above)
Comment (stephan-buckmaster) [reply test/fixtures/language_models.yml:177]: See above comment, it looks like a stretch to combine test-user configuration and actual prod usage values.

File: test/models/api_service_test.rb (hunk 26 above)
Comment (stephan-buckmaster) [reply test/models/api_service_test.rb:5]: Applied

File: app/services/ai_backend/anthropic.rb (hunk 4 above)
Comment (johndoe929) [reply app/services/ai_backend/anthropic.rb:17]: Yes, that's right. So the Assistant > LanguageModel > APIService stack is all per-user. (Maybe some future "teams" features could cause us to revisit that with a more sophisticated permissions model...)

Comment (krschacht) [reply app/services/ai_backend/anthropic.rb:17]: Yes, that's right. So the Assistant > LanguageModel > APIService stack is all per-user. (Maybe some future "teams" features could cause us to revisit that with a more sophisticated permissions model...)

File: test/fixtures/api_services.yml (hunk 24 above)
Comment (krschacht) [reply test/fixtures/api_services.yml:27]: Okay, yes, I think you're right. Maybe it was a mistake that I did that for LanguageModel, but anyway, that's out of scope so I can revisit that later.

At the very least, there is some difference between stuff we're creating which is **connected to a User**, like Assistants and API Services.

File: app/models/user.rb (hunk 22 above)
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/user.rb]: Ok made change, there's probably a gem for this. Look one up?

**Comment by stephan-buckmaster** [comment]
Ok got through more feedback. Now what I see outstanding is to create all records on the user level, including APIService records for OpenAI/Anthropic with their  API URL's.

**Comment by krschacht** [comment]
I’m not sure we need a gem for this. But I’m not sure what symptom we’re
trying to address with this. This was in the other PR and I managed to
eliminate it. But I may be missing something.

**Comment by krschacht** [comment]
I think that’s right. That and the questions on that destroy logic, I’m
still not clear if we need any of that.

File: app/models/user.rb (hunk 22 above)
[the remark this answers is not in the document]
Comment (stephan-buckmaster) [reply app/models/user.rb]: Applied, see last commit, f0190b0

**Comment by krschacht** [comment]
Oh, sorry, earlier I was commenting from my phone and I missed you had already done the delete! thing. That looks good! So yes, it’s just this APIService’s change and then we can merge in!

**Comment by stephan-buckmaster** [comment]
Am I going too far? If all Language Models are assigned to the user, and all users will have copy of the standard API Services, then these APIService records should have the openai-key / anthropic-key. So these two keys should be copied from the users records to the new api_services record -- and the columns should be removed from users table. I think people will get confused if there's so many different places for configuring.

How does this fit in with the sign-up sequence?

**Comment by krschacht** [comment]
@stephan-buckmaster dang, you’re right. This won’t hurt the signup sequence. Currently signup completely ignores the API keys and when you try to chat with an assistant the very first time then it responds with some hard-coded text instructions. Those will be easy to tweak.

**Comment by stephan-buckmaster** [comment]
All right, so I'll work out this reorganization. At least existing users should see no difference.

**Comment by krschacht** [comment]
👍 should we merge in this PR and do that one as a new one?

**Comment by stephan-buckmaster** [comment]
I may have it done in a few hours. But if you don't see a commit -- feel free to merge.

**Comment by krschacht** [comment]
I got a bit bogged down on my PR so I haven’t gotten a chance to circle back on this yet.

**Comment by stephan-buckmaster** [comment]
I have pushed what I have now except for moving to user-managed language model records only. That's because it really is a larger change, it may not be quick review. I made a third pull-request for that, https://github.com/AllYourBot/hostedgpt/pull/425

**Comment by krschacht** [comment]
@stephan-buckmaster I did a full pass through the PR today and I think it's ready to merge in. I made some changes locally as I was reviewing and just pushed up a big commit that blends them all. Here are my notes, just in case you're curious:

* Made controllers actions all redirect back to index
* Changed how settings nav links are rendered (using the settings/application_controller ivar)
* Remove the custom .delete! methods since we have a built-in .deleted! one already
* Renamed language_models.description to language_models.name
* Added back the API key instructions to the new form
* Changed api_service.driver to be an enum field (active record gives you a .keys method for free to get an array of all possibilities)
* Moved the create_without_validations back to the migration since it's really a one-time thing
* Updated a few places we need to use .effective_token rather than .token

Oh, and you had a driver called "Built On" on the API Services, what was that? I think the user should have to select openai or anthropic, right? I removed it but I can revert if I'm mistaken.

I really think it's ready for me to click "merge", but just since I've been staring at this for a few hours I'll do a user-facing run-through one more time tomorrow before and then click merge in the morning. I don't like to merge into main before calling it a night. :)

**Review by krschacht — APPROVED** [review]

**Review by krschacht — APPROVED** [review]

**Comment by krschacht** [comment]
@stephan-buckmaster And it's merged! Whew. This project was quite a big lift.

Today I tried adding Llama 3 myself, through Groq, and fixed a couple small bugs with that. It works so well that I went ahead and made Llama3 a new default within the app and pre-created Groq for people (alongside OpenAI and Anthropic).

A few loose ends that I want to capture while it's top of mind:

1. I'm going to do some more UI work on the settings section of the app
2. When creating a new language model, you had a driver called "built on". Do we need this option? I removed it but I want to confirm that I didn't break something.
3. You left a comment in another thread about adding new APIs via migrations. I agree this feels quite hacky. Do you have any better ideas? I just did it again with the Llama 3 addition.
4. Within this PR we just merged in I went ahead and left the `self.create_without_validation` method in LanguageModel. I saw you fixed up a couple bugs yesterday after my big commit so thank you for that. Could we move it back into the migration? I didn't quite understand why it broke when I tried to move it back there.

**Comment by krschacht** [comment]
@stephan-buckmaster You might appreciate hearing this: after upgrading my personal server to this latest PR, my APIs stopped working. I checked the API keys and they were copied over incorrectly. `user.openai_key` got copied over as `{"p":"Qhq8i6vTyEoEUzqwmcbBQkj6ut/LUvZhdPbIQvR7wcmYq0cMKIshZFN4ufHwSR9KxntMnMBNbYrPGUXxRMU8IZb8iinEH6lhBBN7r0y46kmCDfktc6uc1sE+GLSjBkf6cfSgUJAJz6v+2K9u","h":{"iv":"AOUUAmbvsuDZMUFJ","at":"gHCXP8+k0tpyaoRuPWtzKA=="}}`

I've been digging in for the last couple hours trying to figure out what happened. I realized quickly that this is the raw database value in it's encrypted form (user has encrypts :openai_key) and I got stuck thinking this was a migration issue, like maybe DB encryption wasn't enabled in migrations because of a rails initialization issue.

Finally, I realized that we removed `encrypts :openai_key` from the User table in this same PR which does the migration to copy that value over! This was the fix I just committed:

https://github.com/AllYourBot/hostedgpt/commit/4764a7b5a733fbdeaa216c507489579b0c466ca5

Lesson learned! I don't have a ton of experience thinking about encrypted database contents so this was a new one to me.

One reflection I've had on this whole PR is that it probably would have been worth us merging it in as several PRs along the way. However, the challenge with that is that all of the pieces were really tightly coupled together and the whole architecture was a bit in flux at each stage of the game. Maybe once we had really settled on the scope and had a big PR with lots of work, it probably would have been worth breaking off pieces at a time and landing them. One PR could have just added the models and populated the data, without all the views and the integration with AI backend. Then another one could update the views. Etc. Not a big deal, we got it done either way, but just an interesting reflection.

**Comment by stephan-buckmaster** [comment]
1. Ok
2. About "Built-in." That is obsolete because we assigned them directly now
3. Ok, not something I am familiar with
4.  I feel users can configure their own, just need to know what to do when a new model comes out. They have so many options now. E.g. They can have "OpenAI cheap" "OpenAI premium" .... then when prices change assign what works.
5. Sure, it can go back, it is used in two migrations though, so doesn't look good
 
About the encryption -- that is quite something, I tried it out, and found now problem, and then deleted the "encrypts" on User. 

The feature seemed easier to begin with, then became larger. It might be good to plan features in more detail prior to implemenation. 

One nice thing to add here would be a "check connection" button for API Services, because that can be  difficult to know from the outside.

**Comment by krschacht** [comment]
Great! I handled the migration cleanup. I need to think more about how to help people upgrade/add new LLM models. I would like it to be a little more automated than that since not everyone keeps on top of every AI's companies latest announcements like we probably do. :) I've been thinking about some other mechanism within the app to announce new features so maybe it's related to that. I may setup some managed endpoint that I run, which every instance of the open source app can hit in order to get "announcements" which get displayed in some small, clean way in the UI. I'll keep thinking.

Good point about the "check connection" button.

I went ahead and created tasks for all of these things so I can keep track things:
https://github.com/AllYourBot/hostedgpt/issues?q=sort%3Aupdated-desc+is%3Aopen+label%3A%22LLM+foundation%22

I'm probably going to tackle function/tool calling next since I've been actively working in that area anyway. If you want to take on any of these, just let me know! I know you're traveling and have contributed a lot so no pressure, I just wanted to keep you in the loop.
